compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp

```diff
@@ -236,107 +227,332 @@ static bool setPipeliningMarkers(scf::ForOp forOp, bool pipelineStoreStage) {
+                                        &op, forOp.getBody());
+          }
         }
       }
-      // If not prefected go in the last stage.
-      stageCompute.push_back(&op);
     }
   }
 
-  // Return an empty schedule if the loop is not a candidate to be pipelined.
-  if (loopCanBePipelined || stageCopyToSharedMemory.empty() ||
-      stageCompute.empty())
-    return;
+  // Returns the number of kgroups in the Warp-level MMA operations.
+  int getNumberOfKgroups() { return warpOperations.size(); }
+};
 
-  // Track dependencies of stage 0 ops.
-  llvm::SmallDenseSet<Operation*> deps;
-  llvm::SmallDenseSet<Operation*> stageCopyToSharedMemoryDeps;
-  llvm::SmallDenseSet<Operation*> stageNMinusOneDeps;
-  llvm::SmallDenseSet<Operation*> stageNDeps;
-  for (Operation* op : stageCopyToSharedMemory) {
-    addOpsAndDeps(deps, stageCopyToSharedMemoryDeps, op, forOp.getBody());
+/// Prints the given `funcOp` after a leading `step` comment header.
+void debugMainloopSchedule(MainLoopInfo& mainloop, int numStages,
+                           std::vector<std::pair<Operation*, unsigned>>& ops) {
+  LLVM_DEBUG({
+    llvm::dbgs() << "//--- Mainloop schedule generated for Nvidia Ampere "
+                    "mma.sync TensorCore Pipeline. ---//\n";
+    llvm::dbgs() << " Number of stages: " << numStages << "\n";
+    llvm::dbgs() << " Number of kgroups: " << mainloop.getNumberOfKgroups()
+                 << "\n";
+    llvm::dbgs() << " Number of mainloop instructions " << ops.size() << "\n";
+    llvm::dbgs() << " Mainloop instructions schedule and stage assignment: ";
+    for (auto& stage_op_pair : ops) {
+      llvm::dbgs() << " Stage (" << stage_op_pair.second << ") , Operation: ";
+      stage_op_pair.first->dump();
+    }
+    llvm::dbgs() << "\n\n";
+  });
+}
+
+/// This function returns an *coarse-grained* stage assignment for software
+/// pipelining of the mainloop and a *fine-grained* instruction interleaving.
+/// The schedule provides good performance on Nvidia Ampere architecture using
+/// Ampere-style multi-staged pipeline.
+///
+/// @param forOp the mainloop to pipeline and schedule.
+/// @param ops a vector of pairs: [(operations, pipeline_stage)].
+/// @param numStages the total number of pipeline stages used for multi-buffer.
+static void getNvidiaAmpereTensorCorePipeline(
+    scf::ForOp forOp, std::vector<std::pair<Operation*, unsigned>>& ops,
+    unsigned numStages) {
+  // Analyze the main loop and obtain information for coarse-grained pipelining
+  // and fine-grained instruction scheduling.
+  MainLoopInfo mainloop(forOp);
+
+  // If the mainloop is not schedulable, return an empty schedule.
+  if (!mainloop.isSchedulable) return;
+
+  // NVIDIA Ampere Tensor Core multi-staged pipeline requires at least 2 kgroups
+  // and 3 software pipeline stages. If the conditions are not met, return an
+  // empty schedule.
+  int numKgroups = mainloop.getNumberOfKgroups();
+  if (numKgroups < 2 || numStages < 3) {
```

Comment:
we should be able to support `numStages == 2`? Would the code not work?

Reply:
https://github.com/openxla/iree/pull/12550#discussion_r1129931689

> We need 3 or more stages for Ampere-style multi-staged pipeline: 0th stage for `cp.async`, last stage for `mma.sync` and some `ld.shared`/`ldmatrix`, one before the last stage for load to first kgroup. 

We schedule in stage 0, numStages - 1, and numStages - 2. If this condition is not met, then ops for 0th stage and numStages - 2 goes to the same stage instead of two different stage. I don't know what effect it has on functionality and performance. We can try it out when and if we need it. However, on any Ampere architecture, it won't make sense to have lower than 3 stages.